{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3c4fea02-09b4-4e9e-9fc6-7006e411169b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[{\"collection_id\":\"4d_lung\"},{\"collection_id\":\"acrin_6698\"},{\"collection_id\":\"acrin_contralateral_breast_mr\"},{\"collection_id\":\"acrin_flt_breast\"},{\"collection_id\":\"acrin_nsclc_fdg_pet\"},{\"collection_id\":\"adrenal_acc_ki67_seg\"},{\"collection_id\":\"advanced_mri_breast_lesions\"},{\"collection_id\":\"anti_pd_1_lung\"},{\"collection_id\":\"b_mode_and_ceus_liver\"},{\"collection_id\":\"breast_cancer_screening_dbt\"},{\"collection_id\":\"breast_diagnosis\"},{\"collection_id\":\"breast_mri_nact_pilot\"},{\"collection_id\":\"c4kc_kits\"},{\"collection_id\":\"cc_tumor_heterogeneity\"},{\"collection_id\":\"cmb_aml\"},{\"collection_id\":\"cmb_crc\"},{\"collection_id\":\"cmb_gec\"},{\"collection_id\":\"cmb_lca\"},{\"collection_id\":\"cmb_mel\"},{\"collection_id\":\"cmb_mml\"},{\"collection_id\":\"cmb_pca\"},{\"collection_id\":\"cmmd\"},{\"collection_id\":\"colorectal_liver_metastases\"},{\"collection_id\":\"covid_19_ar\"},{\"collection_id\":\"covid_19_ny_sbu\"},{\"collection_id\":\"cptac_aml\"},{\"collection_id\":\"cptac_brca\"},{\"collection_id\":\"cptac_ccrcc\"},{\"collection_id\":\"cptac_cm\"},{\"collection_id\":\"cptac_coad\"},{\"collection_id\":\"cptac_gbm\"},{\"collection_id\":\"cptac_hnscc\"},{\"collection_id\":\"cptac_lscc\"},{\"collection_id\":\"cptac_luad\"},{\"collection_id\":\"cptac_ov\"},{\"collection_id\":\"cptac_pda\"},{\"collection_id\":\"cptac_sar\"},{\"collection_id\":\"cptac_ucec\"},{\"collection_id\":\"ct_colonography\"},{\"collection_id\":\"ct_lymph_nodes\"},{\"collection_id\":\"ct_phantom4radiomics\"},{\"collection_id\":\"ct_vs_pet_ventilation_imaging\"},{\"collection_id\":\"ctpred_sunitinib_pannet\"},{\"collection_id\":\"dro_toolkit\"},{\"collection_id\":\"duke_breast_cancer_mri\"},{\"collection_id\":\"ea1141\"},{\"collection_id\":\"gbm_dsc_mri_dro\"},{\"collection_id\":\"hcc_tace_seg\"},{\"collection_id\":\"htan_hms\"},{\"collection_id\":\"htan_ohsu\"},{\"collection_id\":\"htan_vanderbilt\"},{\"collection_id\":\"htan_wustl\"},{\"collection_id\":\"icdc_glioma\"},{\"collection_id\":\"ispy1\"},{\"collection_id\":\"ispy2\"},{\"collection_id\":\"lctsc\"},{\"collection_id\":\"lidc_idri\"},{\"collection_id\":\"lung_fused_ct_pathology\"},{\"collection_id\":\"lung_pet_ct_dx\"},{\"collection_id\":\"lung_phantom\"},{\"collection_id\":\"lungct_diagnosis\"},{\"collection_id\":\"midrc_ricord_1a\"},{\"collection_id\":\"midrc_ricord_1b\"},{\"collection_id\":\"midrc_ricord_1c\"},{\"collection_id\":\"mouse_astrocytoma\"},{\"collection_id\":\"mouse_mammary\"},{\"collection_id\":\"naf_prostate\"},{\"collection_id\":\"nlm_visible_human_project\"},{\"collection_id\":\"nlst\"},{\"collection_id\":\"nsclc_radiogenomics\"},{\"collection_id\":\"nsclc_radiomics\"},{\"collection_id\":\"nsclc_radiomics_genomics\"},{\"collection_id\":\"nsclc_radiomics_interobserver1\"},{\"collection_id\":\"pancreas_ct\"},{\"collection_id\":\"pancreatic_ct_cbct_seg\"},{\"collection_id\":\"pdmr_292921_168_r\"},{\"collection_id\":\"pdmr_425362_245_t\"},{\"collection_id\":\"pdmr_521955_158_r4\"},{\"collection_id\":\"pdmr_833975_119_r\"},{\"collection_id\":\"pdmr_997537_175_t\"},{\"collection_id\":\"pdmr_bl0293_f563\"},{\"collection_id\":\"pdmr_texture_analysis\"},{\"collection_id\":\"pediatric_ct_seg\"},{\"collection_id\":\"pelvic_reference_data\"},{\"collection_id\":\"phantom_fda\"},{\"collection_id\":\"prostate_3t\"},{\"collection_id\":\"prostate_anatomical_edge_cases\"},{\"collection_id\":\"prostate_diagnosis\"},{\"collection_id\":\"prostate_fused_mri_pathology\"},{\"collection_id\":\"prostate_mri\"},{\"collection_id\":\"prostate_mri_us_biopsy\"},{\"collection_id\":\"prostatex\"},{\"collection_id\":\"pseudo_phi_dicom_data\"},{\"collection_id\":\"qiba_ct_1c\"},{\"collection_id\":\"qin_breast\"},{\"collection_id\":\"qin_breast_dce_mri\"},{\"collection_id\":\"qin_lung_ct\"},{\"collection_id\":\"qin_prostate_repeatability\"},{\"collection_id\":\"remind\"},{\"collection_id\":\"rider_breast_mri\"},{\"collection_id\":\"rider_lung_ct\"},{\"collection_id\":\"rider_lung_pet_ct\"},{\"collection_id\":\"rider_pilot\"},{\"collection_id\":\"rms_mutation_prediction\"},{\"collection_id\":\"soft_tissue_sarcoma\"},{\"collection_id\":\"spie_aapm_lung_ct_challenge\"},{\"collection_id\":\"stageii_colorectal_ct\"},{\"collection_id\":\"tcga_acc\"},{\"collection_id\":\"tcga_blca\"},{\"collection_id\":\"tcga_brca\"},{\"collection_id\":\"tcga_cesc\"},{\"collection_id\":\"tcga_chol\"},{\"collection_id\":\"tcga_coad\"},{\"collection_id\":\"tcga_dlbc\"},{\"collection_id\":\"tcga_esca\"},{\"collection_id\":\"tcga_gbm\"},{\"collection_id\":\"tcga_hnsc\"},{\"collection_id\":\"tcga_kich\"},{\"collection_id\":\"tcga_kirc\"},{\"collection_id\":\"tcga_kirp\"},{\"collection_id\":\"tcga_lgg\"},{\"collection_id\":\"tcga_lihc\"},{\"collection_id\":\"tcga_luad\"},{\"collection_id\":\"tcga_lusc\"},{\"collection_id\":\"tcga_meso\"},{\"collection_id\":\"tcga_ov\"},{\"collection_id\":\"tcga_paad\"},{\"collection_id\":\"tcga_pcpg\"},{\"collection_id\":\"tcga_prad\"},{\"collection_id\":\"tcga_read\"},{\"collection_id\":\"tcga_sarc\"},{\"collection_id\":\"tcga_skcm\"},{\"collection_id\":\"tcga_stad\"},{\"collection_id\":\"tcga_tgct\"},{\"collection_id\":\"tcga_thca\"},{\"collection_id\":\"tcga_thym\"},{\"collection_id\":\"tcga_ucec\"},{\"collection_id\":\"tcga_ucs\"},{\"collection_id\":\"tcga_uvm\"},{\"collection_id\":\"upenn_gbm\"},{\"collection_id\":\"vestibular_schwannoma_mc_rc\"},{\"collection_id\":\"vestibular_schwannoma_seg\"},{\"collection_id\":\"victre\"}]'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from idc_index import index\n",
    "\n",
    "client = index.IDCClient()\n",
    "query = f\"\"\"\n",
    "select collection_id\n",
    "from index\n",
    "group by collection_id\n",
    "order by collection_id asc\n",
    "\"\"\"\n",
    "df = client.sql_query(query)\n",
    "result = df.to_json(orient=\"records\")\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fc0db53c-1dd1-467f-98c2-1215103a05aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><tr><th>collection_id</th></tr><tr><td>4d_lung</td></tr><tr><td>acrin_6698</td></tr><tr><td>acrin_contralateral_breast_mr</td></tr><tr><td>acrin_flt_breast</td></tr><tr><td>acrin_nsclc_fdg_pet</td></tr><tr><td>adrenal_acc_ki67_seg</td></tr><tr><td>advanced_mri_breast_lesions</td></tr><tr><td>anti_pd_1_lung</td></tr><tr><td>b_mode_and_ceus_liver</td></tr><tr><td>breast_cancer_screening_dbt</td></tr><tr><td>breast_diagnosis</td></tr><tr><td>breast_mri_nact_pilot</td></tr><tr><td>c4kc_kits</td></tr><tr><td>cc_tumor_heterogeneity</td></tr><tr><td>cmb_aml</td></tr><tr><td>cmb_crc</td></tr><tr><td>cmb_gec</td></tr><tr><td>cmb_lca</td></tr><tr><td>cmb_mel</td></tr><tr><td>cmb_mml</td></tr><tr><td>cmb_pca</td></tr><tr><td>cmmd</td></tr><tr><td>colorectal_liver_metastases</td></tr><tr><td>covid_19_ar</td></tr><tr><td>covid_19_ny_sbu</td></tr><tr><td>cptac_aml</td></tr><tr><td>cptac_brca</td></tr><tr><td>cptac_ccrcc</td></tr><tr><td>cptac_cm</td></tr><tr><td>cptac_coad</td></tr><tr><td>cptac_gbm</td></tr><tr><td>cptac_hnscc</td></tr><tr><td>cptac_lscc</td></tr><tr><td>cptac_luad</td></tr><tr><td>cptac_ov</td></tr><tr><td>cptac_pda</td></tr><tr><td>cptac_sar</td></tr><tr><td>cptac_ucec</td></tr><tr><td>ct_colonography</td></tr><tr><td>ct_lymph_nodes</td></tr><tr><td>ct_phantom4radiomics</td></tr><tr><td>ct_vs_pet_ventilation_imaging</td></tr><tr><td>ctpred_sunitinib_pannet</td></tr><tr><td>dro_toolkit</td></tr><tr><td>duke_breast_cancer_mri</td></tr><tr><td>ea1141</td></tr><tr><td>gbm_dsc_mri_dro</td></tr><tr><td>hcc_tace_seg</td></tr><tr><td>htan_hms</td></tr><tr><td>htan_ohsu</td></tr><tr><td>htan_vanderbilt</td></tr><tr><td>htan_wustl</td></tr><tr><td>icdc_glioma</td></tr><tr><td>ispy1</td></tr><tr><td>ispy2</td></tr><tr><td>lctsc</td></tr><tr><td>lidc_idri</td></tr><tr><td>lung_fused_ct_pathology</td></tr><tr><td>lung_pet_ct_dx</td></tr><tr><td>lung_phantom</td></tr><tr><td>lungct_diagnosis</td></tr><tr><td>midrc_ricord_1a</td></tr><tr><td>midrc_ricord_1b</td></tr><tr><td>midrc_ricord_1c</td></tr><tr><td>mouse_astrocytoma</td></tr><tr><td>mouse_mammary</td></tr><tr><td>naf_prostate</td></tr><tr><td>nlm_visible_human_project</td></tr><tr><td>nlst</td></tr><tr><td>nsclc_radiogenomics</td></tr><tr><td>nsclc_radiomics</td></tr><tr><td>nsclc_radiomics_genomics</td></tr><tr><td>nsclc_radiomics_interobserver1</td></tr><tr><td>pancreas_ct</td></tr><tr><td>pancreatic_ct_cbct_seg</td></tr><tr><td>pdmr_292921_168_r</td></tr><tr><td>pdmr_425362_245_t</td></tr><tr><td>pdmr_521955_158_r4</td></tr><tr><td>pdmr_833975_119_r</td></tr><tr><td>pdmr_997537_175_t</td></tr><tr><td>pdmr_bl0293_f563</td></tr><tr><td>pdmr_texture_analysis</td></tr><tr><td>pediatric_ct_seg</td></tr><tr><td>pelvic_reference_data</td></tr><tr><td>phantom_fda</td></tr><tr><td>prostate_3t</td></tr><tr><td>prostate_anatomical_edge_cases</td></tr><tr><td>prostate_diagnosis</td></tr><tr><td>prostate_fused_mri_pathology</td></tr><tr><td>prostate_mri</td></tr><tr><td>prostate_mri_us_biopsy</td></tr><tr><td>prostatex</td></tr><tr><td>pseudo_phi_dicom_data</td></tr><tr><td>qiba_ct_1c</td></tr><tr><td>qin_breast</td></tr><tr><td>qin_breast_dce_mri</td></tr><tr><td>qin_lung_ct</td></tr><tr><td>qin_prostate_repeatability</td></tr><tr><td>remind</td></tr><tr><td>rider_breast_mri</td></tr><tr><td>rider_lung_ct</td></tr><tr><td>rider_lung_pet_ct</td></tr><tr><td>rider_pilot</td></tr><tr><td>rms_mutation_prediction</td></tr><tr><td>soft_tissue_sarcoma</td></tr><tr><td>spie_aapm_lung_ct_challenge</td></tr><tr><td>stageii_colorectal_ct</td></tr><tr><td>tcga_acc</td></tr><tr><td>tcga_blca</td></tr><tr><td>tcga_brca</td></tr><tr><td>tcga_cesc</td></tr><tr><td>tcga_chol</td></tr><tr><td>tcga_coad</td></tr><tr><td>tcga_dlbc</td></tr><tr><td>tcga_esca</td></tr><tr><td>tcga_gbm</td></tr><tr><td>tcga_hnsc</td></tr><tr><td>tcga_kich</td></tr><tr><td>tcga_kirc</td></tr><tr><td>tcga_kirp</td></tr><tr><td>tcga_lgg</td></tr><tr><td>tcga_lihc</td></tr><tr><td>tcga_luad</td></tr><tr><td>tcga_lusc</td></tr><tr><td>tcga_meso</td></tr><tr><td>tcga_ov</td></tr><tr><td>tcga_paad</td></tr><tr><td>tcga_pcpg</td></tr><tr><td>tcga_prad</td></tr><tr><td>tcga_read</td></tr><tr><td>tcga_sarc</td></tr><tr><td>tcga_skcm</td></tr><tr><td>tcga_stad</td></tr><tr><td>tcga_tgct</td></tr><tr><td>tcga_thca</td></tr><tr><td>tcga_thym</td></tr><tr><td>tcga_ucec</td></tr><tr><td>tcga_ucs</td></tr><tr><td>tcga_uvm</td></tr><tr><td>upenn_gbm</td></tr><tr><td>vestibular_schwannoma_mc_rc</td></tr><tr><td>vestibular_schwannoma_seg</td></tr><tr><td>victre</td></tr></table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import json\n",
    "from IPython.display import HTML, display\n",
    "\n",
    "data = json.loads(result)\n",
    "\n",
    "# Convert JSON to HTML table\n",
    "html_table = \"<table><tr><th>collection_id</th></tr>\"\n",
    "for record in data:\n",
    "    html_table += \"<tr><td>{}</td></tr>\".format(record['collection_id'])\n",
    "html_table += \"</table>\"\n",
    "\n",
    "# Display HTML table in Jupyter Notebook\n",
    "display(HTML(html_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8579970c-c134-4fdf-a20e-26c79d375efe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['collection_id', 'analysis_result_id', 'PatientID', 'SeriesInstanceUID',\n",
       "       'StudyInstanceUID', 'source_DOI', 'PatientAge', 'PatientSex',\n",
       "       'StudyDate', 'StudyDescription', 'BodyPartExamined', 'Modality',\n",
       "       'Manufacturer', 'ManufacturerModelName', 'SeriesDate',\n",
       "       'SeriesDescription', 'SeriesNumber', 'instanceCount',\n",
       "       'license_short_name', 'series_aws_url', 'series_size_MB'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from idc_index import index\n",
    "\n",
    "client = index.IDCClient()\n",
    "query = f\"\"\"\n",
    "select *\n",
    "from index\n",
    "\"\"\"\n",
    "#query = f\"\"\"\n",
    "#select StudyInstanceUID, SeriesInstanceUID, series_aws_url, collection_id, PatientID, StudyDate, PatientAge, \"Modality\", \"Manufacturer\", \"ManufacturerModelName\", collection_id, \"StudyDate\", \"StudyDescription\", \"SeriesDescription\", collection_id\n",
    "#from index\n",
    "#\"\"\"\n",
    "df = client.sql_query(query)\n",
    "# fields not found in idc index: ImageType, Patient Weight, SliceThickness, PixelSpacing, tcia_species, tcia_tumorLocation\n",
    "result = df.columns\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "10d62104-0860-43bb-a6c3-9142091b8784",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def get_collections():\n",
    "    client = index.IDCClient()\n",
    "    query = f\"\"\"\n",
    "    select collection_id\n",
    "    from index\n",
    "    group by collection_id\n",
    "    order by collection_id asc\n",
    "    \"\"\"\n",
    "    df = client.sql_query(query)\n",
    "    #result = df.to_json(orient=\"records\")\n",
    "    #result = df.values.tolist()\n",
    "    result_list = [item for sublist in df.values.tolist() for item in sublist]\n",
    "    #result_string = ','.join(result_list)\n",
    "    return result_list\n",
    "\n",
    "def get_patients(collection, filters):\n",
    "    client = index.IDCClient()\n",
    "    \n",
    "    query = f\"\"\"\n",
    "    select PatientID\n",
    "    from index\n",
    "    where collection_id = '{collection}'\n",
    "    \"\"\"\n",
    "    \n",
    "    df = client.sql_query(query)\n",
    "    df_unique = df.drop_duplicates()\n",
    "    patient_ids = df_unique['PatientID'].tolist()\n",
    "\n",
    "    if filters:\n",
    "        filtered_patient_ids = filter_primary_sites(patient_ids, filters)\n",
    "        # Update patient_ids to only include filtered IDs\n",
    "        patient_ids = \"\\\"\" + \"\\\",\\\"\".join(patient_ids) + \"\\\"\"\n",
    "        #patient_ids = filtered_patient_ids.split(',') if filtered_patient_ids else []\n",
    "\n",
    "    result = jsonify(patient_ids)\n",
    "    return result\n",
    "\n",
    "# $patient_ids = \"TCGA-EY-A214,TCGA-02-0003\"\n",
    "# $primary_sites = \"Brain\"\n",
    "# $results = \"TCGA-02-0003\" because TCGA-02-0003 has primary sites Brain but TCGA-EY-A214\n",
    "# curl -sX GET \"http://index:5000/patients/tcga_ov/Brain\"\n",
    "def filter_primary_sites(patient_ids, primary_sites):\n",
    "    if primary_sites is None or primary_sites == \"\":\n",
    "        return patient_ids\n",
    "    \n",
    "    include_or_exclude = \"in\"   \n",
    "    #include_or_exclude = \"ex\"\n",
    "\n",
    "    patient_ids = \"\\\"\" + \"\\\",\\\"\".join(patient_ids.split(',')) + \"\\\"\"\n",
    "    primary_sites = \"\\\"\" + \"\\\",\\\"\".join(primary_sites.split(',')) + \"\\\"\"\n",
    "\n",
    "    command = f\"curl -s --data 'filters={{\\\"op\\\":\\\"and\\\",\\\"content\\\":[{{\\\"op\\\":\\\"in\\\",\\\"content\\\":{{\\\"field\\\":\\\"cases.submitter_id\\\",\\\"value\\\":[{patient_ids}]}}}},{{\\\"op\\\":\\\"{include_or_exclude}\\\",\\\"content\\\":{{\\\"field\\\":\\\"cases.primary_site\\\",\\\"value\\\":[{primary_sites}]}}}}]}}&fields=case_id,submitter_id,primary_site&size=10' 'https://api.gdc.cancer.gov/cases' | jq -r '.data.hits | map(.submitter_id) | join(\\\",\\\")'\"\n",
    "    \n",
    "    output = subprocess.check_output(command, shell=True, text=True, stderr=subprocess.STDOUT)\n",
    "    return output.strip()\n",
    "\n",
    "def filter_experimental_strategies(patient_ids, experimental_strategies):\n",
    "    if experimental_strategies is None or experimental_strategies == \"\":\n",
    "        return patient_ids\n",
    "        \n",
    "    patient_ids = \"\\\"\" + \"\\\",\\\"\".join(patient_ids.split(',')) + \"\\\"\"\n",
    "    experimental_strategies = \"\\\"\" + \"\\\",\\\"\".join(experimental_strategies.split(',')) + \"\\\"\"\n",
    "\n",
    "    command = f\"curl -s --data 'filters={{\\\"op\\\":\\\"and\\\",\\\"content\\\":[{{\\\"op\\\":\\\"in\\\",\\\"content\\\":{{\\\"field\\\":\\\"cases.submitter_id\\\",\\\"value\\\":[{patient_ids}]}}}}]}}&fields=files.experimental_strategy,submitter_id&size=10' https://api.gdc.cancer.gov/cases| jq -r '.data.hits[] | select(.files[].experimental_strategy | IN({experimental_strategies})) | .submitter_id' | sort -u\"\n",
    "\n",
    "    output = subprocess.check_output(command, shell=True, text=True, stderr=subprocess.STDOUT)\n",
    "    return output.strip()\n",
    "\n",
    "# if collections is None:\n",
    "#     collections = get_collections()\n",
    "collections = get_collections()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c9b2b848-660f-4d82-b946-46e6d52d9890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'4d_lung,acrin_6698,acrin_contralateral_breast_mr,acrin_flt_breast,acrin_nsclc_fdg_pet,adrenal_acc_ki67_seg,advanced_mri_breast_lesions,anti_pd_1_lung,b_mode_and_ceus_liver,breast_cancer_screening_dbt,breast_diagnosis,breast_mri_nact_pilot,c4kc_kits,cc_tumor_heterogeneity,cmb_aml,cmb_crc,cmb_gec,cmb_lca,cmb_mel,cmb_mml,cmb_pca,cmmd,colorectal_liver_metastases,covid_19_ar,covid_19_ny_sbu,cptac_aml,cptac_brca,cptac_ccrcc,cptac_cm,cptac_coad,cptac_gbm,cptac_hnscc,cptac_lscc,cptac_luad,cptac_ov,cptac_pda,cptac_sar,cptac_ucec,ct_colonography,ct_lymph_nodes,ct_phantom4radiomics,ct_vs_pet_ventilation_imaging,ctpred_sunitinib_pannet,dro_toolkit,duke_breast_cancer_mri,ea1141,gbm_dsc_mri_dro,hcc_tace_seg,htan_hms,htan_ohsu,htan_vanderbilt,htan_wustl,icdc_glioma,ispy1,ispy2,lctsc,lidc_idri,lung_fused_ct_pathology,lung_pet_ct_dx,lung_phantom,lungct_diagnosis,midrc_ricord_1a,midrc_ricord_1b,midrc_ricord_1c,mouse_astrocytoma,mouse_mammary,naf_prostate,nlm_visible_human_project,nlst,nsclc_radiogenomics,nsclc_radiomics,nsclc_radiomics_genomics,nsclc_radiomics_interobserver1,pancreas_ct,pancreatic_ct_cbct_seg,pdmr_292921_168_r,pdmr_425362_245_t,pdmr_521955_158_r4,pdmr_833975_119_r,pdmr_997537_175_t,pdmr_bl0293_f563,pdmr_texture_analysis,pediatric_ct_seg,pelvic_reference_data,phantom_fda,prostate_3t,prostate_anatomical_edge_cases,prostate_diagnosis,prostate_fused_mri_pathology,prostate_mri,prostate_mri_us_biopsy,prostatex,pseudo_phi_dicom_data,qiba_ct_1c,qin_breast,qin_breast_dce_mri,qin_lung_ct,qin_prostate_repeatability,remind,rider_breast_mri,rider_lung_ct,rider_lung_pet_ct,rider_pilot,rms_mutation_prediction,soft_tissue_sarcoma,spie_aapm_lung_ct_challenge,stageii_colorectal_ct,tcga_acc,tcga_blca,tcga_brca,tcga_cesc,tcga_chol,tcga_coad,tcga_dlbc,tcga_esca,tcga_gbm,tcga_hnsc,tcga_kich,tcga_kirc,tcga_kirp,tcga_lgg,tcga_lihc,tcga_luad,tcga_lusc,tcga_meso,tcga_ov,tcga_paad,tcga_pcpg,tcga_prad,tcga_read,tcga_sarc,tcga_skcm,tcga_stad,tcga_tgct,tcga_thca,tcga_thym,tcga_ucec,tcga_ucs,tcga_uvm,upenn_gbm,vestibular_schwannoma_mc_rc,vestibular_schwannoma_seg,victre'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "collections_str = ','.join(collections)\n",
    "collections_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "26671825-b7ea-4b2c-9a73-a8036465b8c8",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "get_patients() missing 1 required positional argument: 'filters'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[24], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m patients \u001b[38;5;241m=\u001b[39m \u001b[43mget_patients\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mTCGA-GBM\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m patients_str \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(patients)\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# #print(patients_str)\u001b[39;00m\n",
      "\u001b[0;31mTypeError\u001b[0m: get_patients() missing 1 required positional argument: 'filters'"
     ]
    }
   ],
   "source": [
    "patients = get_patients('TCGA-GBM')\n",
    "patients_str = ','.join(patients)\n",
    "# #print(patients_str)\n",
    "print(len(patients))\n",
    "\n",
    "# filtered_patients = filter_primary_sites(patients_str, \"Breast\")\n",
    "filtered_patients_str = filter_primary_sites(patients_str, \"Brain\")\n",
    "filtered_patients = filtered_patients_str.split(',')\n",
    "print(len(filtered_patients))\n",
    "print(filtered_patients)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "140c3ac6-7276-4f07-8c67-b22be7e5cc67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TCGA-BH-A0GY'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filtered_patients = filter_primary_sites(patients_str, 'Breast')\n",
    "filtered_patients = filter_primary_sites(\"TCGA-EY-A214,TCGA-BH-A0GY,TCGA-02-0003\", \"Breast\")\n",
    "#filtered_patients = filter_primary_sites(\"TCGA-EY-A214,TCGA-BH-A0GY,TCGA-02-0003\", \"\")\n",
    "filtered_patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a8691b3-b33c-4743-abd5-99b70f94ce1b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'filtered_patients' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m key \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPatientID\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      4\u001b[0m where \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwhere 1 = 1\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 5\u001b[0m values \u001b[38;5;241m=\u001b[39m \u001b[43mfiltered_patients\u001b[49m\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(values) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m      7\u001b[0m     where \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m and \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m key \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m in (\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(values) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'filtered_patients' is not defined"
     ]
    }
   ],
   "source": [
    "#select = 'select PatientID from index '\n",
    "select = 'select * from index '\n",
    "key = 'PatientID'\n",
    "where = 'where 1 = 1'\n",
    "values = filtered_patients.split(',')\n",
    "if len(values) > 0:\n",
    "    where += \" and \" + key + \" in ('\" + \"','\".join(values) + \"')\"\n",
    "else:\n",
    "    where += \" and \" + key + \" = \" + value + \"'\"\n",
    "\n",
    "client = index.IDCClient()\n",
    "query = select + where\n",
    "print(query)\n",
    "df = client.sql_query(query)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "19bdcd95-5056-45c5-9a42-5c9de5b36514",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     s3://idc-open-data/25d0a388-67fe-4fc3-9b93-b17...\n",
       "1     s3://idc-open-data/6a37f4a0-aeda-4b1d-9945-e6d...\n",
       "2     s3://idc-open-data/6ea28924-2058-4933-b72c-aad...\n",
       "3     s3://idc-open-data/1c6eac41-33fc-4372-a201-036...\n",
       "4     s3://idc-open-data/df866683-b180-47db-ac04-c5f...\n",
       "5     s3://idc-open-data/a8a6b4d8-0c0a-4e49-b40e-189...\n",
       "6     s3://idc-open-data/447342ef-c225-49f4-8dd4-659...\n",
       "7     s3://idc-open-data/54d23b21-843d-4931-8ea5-300...\n",
       "8     s3://idc-open-data/7992bbaf-0ba9-4814-96f3-8ff...\n",
       "9     s3://idc-open-data/61ec2800-b4cd-45d4-a5b8-354...\n",
       "10    s3://idc-open-data/9c1215aa-ab97-4480-a523-0d6...\n",
       "11    s3://idc-open-data/f69bb966-fbda-41cc-af4a-f2b...\n",
       "12    s3://idc-open-data/c2ff8d6f-5783-4bcf-a051-39b...\n",
       "13    s3://idc-open-data/1e4939ae-492d-46e9-b3e6-1c6...\n",
       "14    s3://idc-open-data/5d1e2e48-6874-4f5f-b689-473...\n",
       "15    s3://idc-open-data/01a52e2d-a05d-4c00-9f3e-bec...\n",
       "16    s3://idc-open-data/03553c0a-b248-4b00-8ba8-566...\n",
       "Name: series_aws_url, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['series_aws_url']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "af79d14c-ec88-44dd-a30c-ee96f4dd921c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'25d0a388-67fe-4fc3-9b93-b1720b4e7129.tgz'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import shutil\n",
    "\n",
    "def execute_command(command, output=True):\n",
    "    try:\n",
    "        result = subprocess.run(command, shell=True, text=True, capture_output=True, check=True)\n",
    "        result = result.stdout.strip() if result.stdout else \"\"\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        result = e.stderr.strip()\n",
    "    if output:\n",
    "        return result\n",
    "\n",
    "def download_dicom_dir(aws_s3_path):\n",
    "    aws_profile = \"default\"  # Adjust as needed\n",
    "    aws_config_dir = \"/root/.aws/\"\n",
    "    os.environ['AWS_CONFIG_FILE'] = os.path.join(aws_config_dir, 'config')\n",
    "    os.environ['AWS_SHARED_CREDENTIALS_FILE'] = os.path.join(aws_config_dir, 'credentials')\n",
    "\n",
    "    dicom_dir = aws_s3_path.split(\"/\")[-2]\n",
    "    dicom_tar_file = dicom_dir + \".tgz\"\n",
    "\n",
    "    if not os.path.exists(dicom_tar_file):\n",
    "        execute_command(f\"aws s3 s3://idc-open-data/{dicom_dir}/ /app/{dicom_dir} --recursive\", output=False)\n",
    "        execute_command(f\"tar -czvf {dicom_tar_file} /app/{dicom_dir}\")\n",
    "        shutil.rmtree(dicom_dir)\n",
    "\n",
    "    if os.path.exists(dicom_tar_file):\n",
    "        return dicom_tar_file\n",
    "\n",
    "# download dicom dir\n",
    "download_dicom_dir(df['series_aws_url'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fe360a54-8647-42bd-be7f-c280d9ce17a5",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'patients_str' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m key \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPatientID\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      4\u001b[0m where \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwhere 1 = 1\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 5\u001b[0m values \u001b[38;5;241m=\u001b[39m \u001b[43mpatients_str\u001b[49m\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(values) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m      7\u001b[0m     where \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m and \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m key \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m in (\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(values) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'patients_str' is not defined"
     ]
    }
   ],
   "source": [
    "#select = 'select PatientID from index '\n",
    "select = 'select * from index '\n",
    "key = 'PatientID'\n",
    "where = 'where 1 = 1'\n",
    "values = patients_str.split(',')\n",
    "if len(values) > 0:\n",
    "    where += \" and \" + key + \" in ('\" + \"','\".join(values) + \"')\"\n",
    "else:\n",
    "    where += \" and \" + key + \" = \" + value + \"'\"\n",
    "\n",
    "client = index.IDCClient()\n",
    "query = select + where\n",
    "print(query)\n",
    "df = client.sql_query(query)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b6c1dea1-8463-41ef-91cc-9f5f3d861f82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_primary_sites(\"Breast_MRI_001\", \"Breast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4cbb783c-0412-4168-a358-839c8888a908",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TCGA-BH-A0GY'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter_primary_sites(\"TCGA-EY-A214,TCGA-BH-A0GY,TCGA-02-0003\", \"Breast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "325dc34a-37c2-445d-9ff9-9f6ea7ce67b8",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Working outside of application context.\n\nThis typically means that you attempted to use functionality that needed\nthe current application. To solve this, set up an application context\nwith app.app_context(). See the documentation for more information.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 40\u001b[0m\n\u001b[1;32m     37\u001b[0m     result \u001b[38;5;241m=\u001b[39m jsonify(patient_ids)\n\u001b[1;32m     38\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[0;32m---> 40\u001b[0m \u001b[43mget_patients\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtcga_ov\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mBrain\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[14], line 37\u001b[0m, in \u001b[0;36mget_patients\u001b[0;34m(collection, filters)\u001b[0m\n\u001b[1;32m     34\u001b[0m     patient_ids \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(patient_ids)\n\u001b[1;32m     35\u001b[0m     filtered_patient_ids \u001b[38;5;241m=\u001b[39m filter_primary_sites(patient_ids, filters)\n\u001b[0;32m---> 37\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mjsonify\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpatient_ids\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     38\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/flask/json/__init__.py:170\u001b[0m, in \u001b[0;36mjsonify\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    138\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mjsonify\u001b[39m(\u001b[38;5;241m*\u001b[39margs: t\u001b[38;5;241m.\u001b[39mAny, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: t\u001b[38;5;241m.\u001b[39mAny) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Response:\n\u001b[1;32m    139\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Serialize the given arguments as JSON, and return a\u001b[39;00m\n\u001b[1;32m    140\u001b[0m \u001b[38;5;124;03m    :class:`~flask.Response` object with the ``application/json``\u001b[39;00m\n\u001b[1;32m    141\u001b[0m \u001b[38;5;124;03m    mimetype. A dict or list returned from a view will be converted to a\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;124;03m    .. versionadded:: 0.2\u001b[39;00m\n\u001b[1;32m    169\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 170\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcurrent_app\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjson\u001b[49m\u001b[38;5;241m.\u001b[39mresponse(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/werkzeug/local.py:311\u001b[0m, in \u001b[0;36m_ProxyLookup.__get__\u001b[0;34m(self, instance, owner)\u001b[0m\n\u001b[1;32m    308\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 311\u001b[0m     obj \u001b[38;5;241m=\u001b[39m \u001b[43minstance\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_current_object\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    312\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m:\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfallback \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/site-packages/werkzeug/local.py:508\u001b[0m, in \u001b[0;36mLocalProxy.__init__.<locals>._get_current_object\u001b[0;34m()\u001b[0m\n\u001b[1;32m    506\u001b[0m     obj \u001b[38;5;241m=\u001b[39m local\u001b[38;5;241m.\u001b[39mget()\n\u001b[1;32m    507\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mLookupError\u001b[39;00m:\n\u001b[0;32m--> 508\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(unbound_message) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    510\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m get_name(obj)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Working outside of application context.\n\nThis typically means that you attempted to use functionality that needed\nthe current application. To solve this, set up an application context\nwith app.app_context(). See the documentation for more information."
     ]
    }
   ],
   "source": [
    "from idc_index import index\n",
    "import subprocess\n",
    "\n",
    "#from flask import jsonify\n",
    "import json\n",
    "\n",
    "def filter_primary_sites(patient_ids, primary_sites):\n",
    "    if primary_sites is None or primary_sites == \"\":\n",
    "        return patient_ids\n",
    "    \n",
    "    include_or_exclude = \"in\"\n",
    "    #include_or_exclude = \"ex\"\n",
    "\n",
    "    patient_ids = \"\\\"\" + \"\\\",\\\"\".join(patient_ids.split(',')) + \"\\\"\"\n",
    "    primary_sites = \"\\\"\" + \"\\\",\\\"\".join(primary_sites.split(',')) + \"\\\"\"\n",
    "\n",
    "    command = f\"curl -s --data 'filters={{\\\"op\\\":\\\"and\\\",\\\"content\\\":[{{\\\"op\\\":\\\"in\\\",\\\"content\\\":{{\\\"field\\\":\\\"cases.submitter_id\\\",\\\"value\\\":[{patient_ids}]}}}},{{\\\"op\\\":\\\"{include_or_exclude}\\\",\\\"content\\\":{{\\\"field\\\":\\\"cases.primary_site\\\",\\\"value\\\":[{primary_sites}]}}}}]}}&fields=case_id,submitter_id,primary_site&size=10' 'https://api.gdc.cancer.gov/cases' | jq -r '.data.hits | map(.submitter_id) | join(\\\",\\\")'\"\n",
    "    \n",
    "    output = subprocess.check_output(command, shell=True, text=True, stderr=subprocess.STDOUT)\n",
    "    return output.strip()\n",
    "    \n",
    "def get_patients(collection, filters):\n",
    "    client = index.IDCClient()\n",
    "    \n",
    "    query = f\"\"\"\n",
    "    select PatientID\n",
    "    from index\n",
    "    where collection_id = '{collection}'\n",
    "    \"\"\"\n",
    "    \n",
    "    df = client.sql_query(query)\n",
    "    df_unique = df.drop_duplicates()\n",
    "    patient_ids = df_unique['PatientID'].tolist()\n",
    "\n",
    "    if filters:\n",
    "        patient_ids = \",\".join(patient_ids)\n",
    "        filtered_patient_ids = filter_primary_sites(patient_ids, filters)\n",
    "\n",
    "    #result = jsonify(patient_ids) # for flask only\n",
    "    result = json.dumps(data)\n",
    "    return result\n",
    "\n",
    "get_patients(\"tcga_ov\", \"Brain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0b060ea-fa7a-43c3-80cd-bfb67ecd0798",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test 1: No filters\n",
      "Genomic data for patient TCGA-AC-A3W5:\n",
      "UUID: a88c168e-4bba-4bd2-9c0c-77934444cc1c, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 0a40467f-9495-4c5b-b56e-3347a3ee0572, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 950ee56c-df20-4b39-867e-c73f17d0f4f6, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: 62d4515f-a30b-4b1a-b2dd-c8bf9476e803, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: c5039e7e-14d8-4e59-ba20-0d3fa38b86df, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: c10090b8-fa42-4e32-86a0-9dedceb5b605, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: d0c967b5-78ee-4bc4-9493-3cbe44af1bf5, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: 8162d394-8b64-4da2-9f5b-d164c54b9608, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 403b5cef-8173-47c7-b56a-cc94dcfbb2e3, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 1e6b79ff-9787-4cbe-b19d-ebabb6b43589, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 41fd4c6c-2932-4c6b-914f-7abaea9dc8a0, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: 22656b71-476a-4b5f-9f3b-82afb29a34bc, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: edbb5de2-d0b5-4a07-a0e1-19e952a5117d, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: a9baecf0-5549-4396-8805-a6d1681d11cd, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 43569ab9-5a32-45ef-a07e-e93f45416a16, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: a24d7574-e419-47ce-9293-b9b29f854a10, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: 22f38342-3f1a-47e5-9025-c19719fa21d0, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: f434b393-9e32-4ca2-aa1a-3b0de36b5ffe, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: 5907be67-69f6-4339-8264-952a2462bea5, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: b5fd0411-7565-4801-874b-6c88b13733b6, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: c15d7887-3e27-4456-bd6b-34bf48a9db4e, Data Type: Raw Simple Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: 38406d21-f213-4d98-b45f-a8516478edb5, Data Type: Masked Copy Number Segment, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: a15f37c5-1b76-44d3-9fb2-2755d90d6562, Data Type: Slide Image, Data Category: Biospecimen, Experimental Strategy: Tissue Slide\n",
      "UUID: 5e863e9f-1449-4a38-ac00-cfd7c21ba6e2, Data Type: Gene Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "UUID: 902542d4-5653-4883-97f7-923500a712ee, Data Type: Gene Level Copy Number, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 94a961a8-af2b-4da3-a58f-8fa05dc8d8fb, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: bb733e98-3a54-47c8-b08e-622bac96a62f, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: RNA-Seq\n",
      "UUID: cd417f97-ab71-4f0a-8e3a-0b9077f94638, Data Type: Simple Germline Variation, Data Category: Simple Nucleotide Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 24371099-70d6-4869-aa83-04d0fed53f42, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: WGS\n",
      "UUID: 0bdedc35-7f54-45e9-b18d-198c1ef4694e, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: aa8ade7e-f57c-4ba6-88d5-7e1e3d341443, Data Type: Isoform Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: miRNA-Seq\n",
      "UUID: 9425fe04-6b2f-4adc-aa53-b0cfeb74cebf, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: WXS\n",
      "UUID: 2727f6d7-4f4b-4651-9807-3312c52befd3, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: RNA-Seq\n",
      "UUID: 82c4d5ae-ea1b-45d5-aedf-609edf0bcbd4, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: f7f48058-1476-46e6-88c0-b313a12b59bd, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: WXS\n",
      "UUID: ebd13509-602b-4e8e-aad8-9be5cae2890c, Data Type: Simple Germline Variation, Data Category: Simple Nucleotide Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: bf3fd5be-f72c-4edb-966e-a1dcf5395963, Data Type: Splice Junction Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "UUID: 1675394d-373f-4f64-996a-8d847c648caf, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: a93315ad-063e-443b-baec-6368d1c37785, Data Type: Clinical Supplement, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: c71acc1f-5752-4cc3-99f4-46133befea86, Data Type: Raw Intensities, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 3e63c528-6403-4781-8555-2daaf4b0d1dd, Data Type: Allele-specific Copy Number Segment, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 08a63743-09a2-4f30-8948-765c45c64aa7, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: miRNA-Seq\n",
      "UUID: aa931745-919a-4687-82ed-34c4d1e53fe1, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: 0fa12edb-9b16-4ba2-ae3c-929b93d5dcb8, Data Type: Slide Image, Data Category: Biospecimen, Experimental Strategy: Diagnostic Slide\n",
      "UUID: 8f7ed391-4e7b-48e2-8f0d-b441c837fb61, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: 94a8563f-27a0-4e1f-821b-0d7d0e46bee3, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: RNA-Seq\n",
      "UUID: e06aedec-9ec6-4558-809d-587bff328966, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: 89b4b795-8716-4704-830e-91381919399c, Data Type: Copy Number Segment, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 27109933-254a-4ada-a5d0-556453b79fd3, Data Type: Protein Expression Quantification, Data Category: Proteome Profiling, Experimental Strategy: Reverse Phase Protein Array\n",
      "UUID: 769647dd-c77a-43a0-b9f3-401fec6f84ab, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: f4f690cc-6725-46de-a53b-f6f3c143ea89, Data Type: Raw Simple Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: 61cd84d3-2659-4e92-a1f7-69023377383b, Data Type: Methylation Beta Value, Data Category: DNA Methylation, Experimental Strategy: Methylation Array\n",
      "UUID: 76f936eb-7bca-4bf8-9210-35a6b0c4970d, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: e8a263d7-4230-4bec-92ab-8b80be0000fc, Data Type: Gene Level Copy Number, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: d766dfe0-ea00-479f-8d1d-82ecb98dd2a9, Data Type: Raw Simple Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: f8ebd592-e64d-4387-8704-ca83c5610e35, Data Type: Raw Intensities, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 3547414c-7f4e-49b4-813f-bc1dbf103fe9, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: 95d72a48-88b5-49a4-80aa-6e39f9821979, Data Type: Masked Intensities, Data Category: DNA Methylation, Experimental Strategy: Methylation Array\n",
      "UUID: 50e2d5d6-0678-4db7-b849-df7ccd20d424, Data Type: Raw Simple Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: ebf1e95b-4505-420a-9e2b-c552733ceef6, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: 70c28a05-d777-4f2b-943d-2d5d09ad1c20, Data Type: Annotated Somatic Mutation, Data Category: Simple Nucleotide Variation, Experimental Strategy: WXS\n",
      "UUID: dab7a763-f3c8-4532-acb6-1c72f8f29265, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: d05e1fe7-e3c9-4205-abac-c1f57ccddad6, Data Type: Biospecimen Supplement, Data Category: Biospecimen, Experimental Strategy: N/A\n",
      "UUID: d3480f1c-a301-4799-9d11-afdf81a8bef7, Data Type: Copy Number Segment, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: 3ada67b6-bcdd-4682-b99b-89956219718c, Data Type: miRNA Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: miRNA-Seq\n",
      "UUID: 234a8d5e-2ad6-4ace-b8fc-e575a8903263, Data Type: Masked Intensities, Data Category: DNA Methylation, Experimental Strategy: Methylation Array\n",
      "UUID: 49d506d7-09be-4eaa-b1eb-ff7207bd02e3, Data Type: Allele-specific Copy Number Segment, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: c78d5190-ac60-4af0-a50a-14183e685b9c, Data Type: Gene Level Copy Number, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: fa117dd4-b55e-411a-96d1-0b22eb0913af, Data Type: Masked Copy Number Segment, Data Category: Copy Number Variation, Experimental Strategy: Genotyping Array\n",
      "UUID: be7ad45d-05de-4cc8-83d2-3bebaf45d6bd, Data Type: Pathology Report, Data Category: Clinical, Experimental Strategy: N/A\n",
      "UUID: b6aa44b9-8ac4-4b7e-9bee-56e741c7b7fc, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: WGS\n",
      "\n",
      "Test 2: Filter by data category\n",
      "Genomic data for patient TCGA-AC-A3W5 (Transcriptome Profiling only):\n",
      "UUID: 5e863e9f-1449-4a38-ac00-cfd7c21ba6e2, Data Type: Gene Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "UUID: aa8ade7e-f57c-4ba6-88d5-7e1e3d341443, Data Type: Isoform Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: miRNA-Seq\n",
      "UUID: bf3fd5be-f72c-4edb-966e-a1dcf5395963, Data Type: Splice Junction Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "UUID: 3ada67b6-bcdd-4682-b99b-89956219718c, Data Type: miRNA Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: miRNA-Seq\n",
      "\n",
      "Test 3: Filter by data type\n",
      "Genomic data for patient TCGA-AC-A3W5 (Gene Expression Quantification only):\n",
      "UUID: 5e863e9f-1449-4a38-ac00-cfd7c21ba6e2, Data Type: Gene Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "\n",
      "Test 4: Filter by experimental strategy\n",
      "Genomic data for patient TCGA-AC-A3W5 (RNA-Seq only):\n",
      "UUID: 5e863e9f-1449-4a38-ac00-cfd7c21ba6e2, Data Type: Gene Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "UUID: bb733e98-3a54-47c8-b08e-622bac96a62f, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: RNA-Seq\n",
      "UUID: 2727f6d7-4f4b-4651-9807-3312c52befd3, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: RNA-Seq\n",
      "UUID: bf3fd5be-f72c-4edb-966e-a1dcf5395963, Data Type: Splice Junction Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "UUID: aa931745-919a-4687-82ed-34c4d1e53fe1, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: 94a8563f-27a0-4e1f-821b-0d7d0e46bee3, Data Type: Aligned Reads, Data Category: Sequencing Reads, Experimental Strategy: RNA-Seq\n",
      "UUID: 3547414c-7f4e-49b4-813f-bc1dbf103fe9, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: ebf1e95b-4505-420a-9e2b-c552733ceef6, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "UUID: dab7a763-f3c8-4532-acb6-1c72f8f29265, Data Type: Transcript Fusion, Data Category: Structural Variation, Experimental Strategy: RNA-Seq\n",
      "\n",
      "Test 5: Combining multiple filters\n",
      "Genomic data for patient TCGA-AC-A3W5 (Transcriptome Profiling, Gene Expression Quantification, RNA-Seq):\n",
      "UUID: 5e863e9f-1449-4a38-ac00-cfd7c21ba6e2, Data Type: Gene Expression Quantification, Data Category: Transcriptome Profiling, Experimental Strategy: RNA-Seq\n",
      "\n",
      "Download links for filtered results:\n",
      "  Gene Expression Quantification (RNA-Seq): https://api.gdc.cancer.gov/data/5e863e9f-1449-4a38-ac00-cfd7c21ba6e2\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "def get_genomic_data_uuids(patient_id, data_categories=None, data_types=None, experimental_strategies=None):\n",
    "    \"\"\"\n",
    "    Get genomic data file UUIDs for a given PatientID with optional filters.\n",
    "    \n",
    "    Args:\n",
    "    patient_id (str): The patient ID to query.\n",
    "    data_categories (list): Optional list of data categories to filter by.\n",
    "    data_types (list): Optional list of data types to filter by.\n",
    "    experimental_strategies (list): Optional list of experimental strategies to filter by.\n",
    "    \n",
    "    Returns:\n",
    "    A list of dictionaries, each containing 'uuid', 'data_type', 'data_category', and 'experimental_strategy' for available genomic data.\n",
    "    \"\"\"\n",
    "    base_url = \"https://api.gdc.cancer.gov/files\"\n",
    "    \n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\": [\n",
    "            {\n",
    "                \"op\": \"in\",\n",
    "                \"content\": {\n",
    "                    \"field\": \"cases.submitter_id\",\n",
    "                    \"value\": [patient_id]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    if data_categories:\n",
    "        filters[\"content\"].append({\n",
    "            \"op\": \"in\",\n",
    "            \"content\": {\n",
    "                \"field\": \"data_category\",\n",
    "                \"value\": data_categories\n",
    "            }\n",
    "        })\n",
    "    \n",
    "    if data_types:\n",
    "        filters[\"content\"].append({\n",
    "            \"op\": \"in\",\n",
    "            \"content\": {\n",
    "                \"field\": \"data_type\",\n",
    "                \"value\": data_types\n",
    "            }\n",
    "        })\n",
    "    \n",
    "    if experimental_strategies:\n",
    "        filters[\"content\"].append({\n",
    "            \"op\": \"in\",\n",
    "            \"content\": {\n",
    "                \"field\": \"experimental_strategy\",\n",
    "                \"value\": experimental_strategies\n",
    "            }\n",
    "        })\n",
    "    \n",
    "    params = {\n",
    "        \"filters\": json.dumps(filters),\n",
    "        \"fields\": \"file_id,data_type,data_category,experimental_strategy\",\n",
    "        \"format\": \"JSON\",\n",
    "        \"size\": \"100\"\n",
    "    }\n",
    "    \n",
    "    response = requests.get(base_url, params=params)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        return [{\n",
    "            \"uuid\": hit[\"file_id\"],\n",
    "            \"data_type\": hit[\"data_type\"],\n",
    "            \"data_category\": hit[\"data_category\"],\n",
    "            \"experimental_strategy\": hit.get(\"experimental_strategy\", \"N/A\")\n",
    "        } for hit in data[\"data\"][\"hits\"]]\n",
    "    else:\n",
    "        print(f\"Error: {response.status_code}\")\n",
    "        return []\n",
    "\n",
    "# Test the function with a sample patient ID and various filters\n",
    "sample_patient_id = \"TCGA-AC-A3W5\"  # Replace with a valid patient ID from your dataset\n",
    "\n",
    "print(\"Test 1: No filters\")\n",
    "results = get_genomic_data_uuids(sample_patient_id)\n",
    "print(f\"Genomic data for patient {sample_patient_id}:\")\n",
    "for item in results:\n",
    "    print(f\"UUID: {item['uuid']}, Data Type: {item['data_type']}, Data Category: {item['data_category']}, Experimental Strategy: {item['experimental_strategy']}\")\n",
    "\n",
    "print(\"\\nTest 2: Filter by data category\")\n",
    "results = get_genomic_data_uuids(sample_patient_id, data_categories=[\"Transcriptome Profiling\"])\n",
    "print(f\"Genomic data for patient {sample_patient_id} (Transcriptome Profiling only):\")\n",
    "for item in results:\n",
    "    print(f\"UUID: {item['uuid']}, Data Type: {item['data_type']}, Data Category: {item['data_category']}, Experimental Strategy: {item['experimental_strategy']}\")\n",
    "\n",
    "print(\"\\nTest 3: Filter by data type\")\n",
    "results = get_genomic_data_uuids(sample_patient_id, data_types=[\"Gene Expression Quantification\"])\n",
    "print(f\"Genomic data for patient {sample_patient_id} (Gene Expression Quantification only):\")\n",
    "for item in results:\n",
    "    print(f\"UUID: {item['uuid']}, Data Type: {item['data_type']}, Data Category: {item['data_category']}, Experimental Strategy: {item['experimental_strategy']}\")\n",
    "\n",
    "print(\"\\nTest 4: Filter by experimental strategy\")\n",
    "results = get_genomic_data_uuids(sample_patient_id, experimental_strategies=[\"RNA-Seq\"])\n",
    "print(f\"Genomic data for patient {sample_patient_id} (RNA-Seq only):\")\n",
    "for item in results:\n",
    "    print(f\"UUID: {item['uuid']}, Data Type: {item['data_type']}, Data Category: {item['data_category']}, Experimental Strategy: {item['experimental_strategy']}\")\n",
    "\n",
    "print(\"\\nTest 5: Combining multiple filters\")\n",
    "results = get_genomic_data_uuids(sample_patient_id, \n",
    "                                 data_categories=[\"Transcriptome Profiling\"],\n",
    "                                 data_types=[\"Gene Expression Quantification\"],\n",
    "                                 experimental_strategies=[\"RNA-Seq\"])\n",
    "print(f\"Genomic data for patient {sample_patient_id} (Transcriptome Profiling, Gene Expression Quantification, RNA-Seq):\")\n",
    "for item in results:\n",
    "    print(f\"UUID: {item['uuid']}, Data Type: {item['data_type']}, Data Category: {item['data_category']}, Experimental Strategy: {item['experimental_strategy']}\")\n",
    "\n",
    "# Generate download URLs\n",
    "base_download_url = \"https://api.gdc.cancer.gov/data/\"\n",
    "print(\"\\nDownload links for filtered results:\")\n",
    "for item in results:\n",
    "    download_url = f\"{base_download_url}{item['uuid']}\"\n",
    "    print(f\"  {item['data_type']} ({item['experimental_strategy']}): {download_url}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0951faaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'uuid': 'a88c168e-4bba-4bd2-9c0c-77934444cc1c',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '0a40467f-9495-4c5b-b56e-3347a3ee0572',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '950ee56c-df20-4b39-867e-c73f17d0f4f6',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '62d4515f-a30b-4b1a-b2dd-c8bf9476e803',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'c5039e7e-14d8-4e59-ba20-0d3fa38b86df',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'c10090b8-fa42-4e32-86a0-9dedceb5b605',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'd0c967b5-78ee-4bc4-9493-3cbe44af1bf5',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '8162d394-8b64-4da2-9f5b-d164c54b9608',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '403b5cef-8173-47c7-b56a-cc94dcfbb2e3',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '1e6b79ff-9787-4cbe-b19d-ebabb6b43589',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '41fd4c6c-2932-4c6b-914f-7abaea9dc8a0',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '22656b71-476a-4b5f-9f3b-82afb29a34bc',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'edbb5de2-d0b5-4a07-a0e1-19e952a5117d',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'a9baecf0-5549-4396-8805-a6d1681d11cd',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '43569ab9-5a32-45ef-a07e-e93f45416a16',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'a24d7574-e419-47ce-9293-b9b29f854a10',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '22f38342-3f1a-47e5-9025-c19719fa21d0',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'f434b393-9e32-4ca2-aa1a-3b0de36b5ffe',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '5907be67-69f6-4339-8264-952a2462bea5',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'b5fd0411-7565-4801-874b-6c88b13733b6',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'c15d7887-3e27-4456-bd6b-34bf48a9db4e',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': '38406d21-f213-4d98-b45f-a8516478edb5',\n",
       "  'data_type': 'Masked Copy Number Segment',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': 'a15f37c5-1b76-44d3-9fb2-2755d90d6562',\n",
       "  'data_type': 'Slide Image',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'Tissue Slide'},\n",
       " {'uuid': '5e863e9f-1449-4a38-ac00-cfd7c21ba6e2',\n",
       "  'data_type': 'Gene Expression Quantification',\n",
       "  'data_category': 'Transcriptome Profiling',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': '902542d4-5653-4883-97f7-923500a712ee',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '94a961a8-af2b-4da3-a58f-8fa05dc8d8fb',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'bb733e98-3a54-47c8-b08e-622bac96a62f',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': 'cd417f97-ab71-4f0a-8e3a-0b9077f94638',\n",
       "  'data_type': 'Simple Germline Variation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '24371099-70d6-4869-aa83-04d0fed53f42',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'WGS'},\n",
       " {'uuid': '0bdedc35-7f54-45e9-b18d-198c1ef4694e',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'aa8ade7e-f57c-4ba6-88d5-7e1e3d341443',\n",
       "  'data_type': 'Isoform Expression Quantification',\n",
       "  'data_category': 'Transcriptome Profiling',\n",
       "  'experimental_strategy': 'miRNA-Seq'},\n",
       " {'uuid': '9425fe04-6b2f-4adc-aa53-b0cfeb74cebf',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': '2727f6d7-4f4b-4651-9807-3312c52befd3',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': '82c4d5ae-ea1b-45d5-aedf-609edf0bcbd4',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'f7f48058-1476-46e6-88c0-b313a12b59bd',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'ebd13509-602b-4e8e-aad8-9be5cae2890c',\n",
       "  'data_type': 'Simple Germline Variation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': 'bf3fd5be-f72c-4edb-966e-a1dcf5395963',\n",
       "  'data_type': 'Splice Junction Quantification',\n",
       "  'data_category': 'Transcriptome Profiling',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': '1675394d-373f-4f64-996a-8d847c648caf',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'a93315ad-063e-443b-baec-6368d1c37785',\n",
       "  'data_type': 'Clinical Supplement',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'c71acc1f-5752-4cc3-99f4-46133befea86',\n",
       "  'data_type': 'Raw Intensities',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '3e63c528-6403-4781-8555-2daaf4b0d1dd',\n",
       "  'data_type': 'Allele-specific Copy Number Segment',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '08a63743-09a2-4f30-8948-765c45c64aa7',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'miRNA-Seq'},\n",
       " {'uuid': 'aa931745-919a-4687-82ed-34c4d1e53fe1',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'data_category': 'Structural Variation',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': '0fa12edb-9b16-4ba2-ae3c-929b93d5dcb8',\n",
       "  'data_type': 'Slide Image',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'Diagnostic Slide'},\n",
       " {'uuid': '8f7ed391-4e7b-48e2-8f0d-b441c837fb61',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': '94a8563f-27a0-4e1f-821b-0d7d0e46bee3',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': 'e06aedec-9ec6-4558-809d-587bff328966',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': '89b4b795-8716-4704-830e-91381919399c',\n",
       "  'data_type': 'Copy Number Segment',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '27109933-254a-4ada-a5d0-556453b79fd3',\n",
       "  'data_type': 'Protein Expression Quantification',\n",
       "  'data_category': 'Proteome Profiling',\n",
       "  'experimental_strategy': 'Reverse Phase Protein Array'},\n",
       " {'uuid': '769647dd-c77a-43a0-b9f3-401fec6f84ab',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'f4f690cc-6725-46de-a53b-f6f3c143ea89',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': '61cd84d3-2659-4e92-a1f7-69023377383b',\n",
       "  'data_type': 'Methylation Beta Value',\n",
       "  'data_category': 'DNA Methylation',\n",
       "  'experimental_strategy': 'Methylation Array'},\n",
       " {'uuid': '76f936eb-7bca-4bf8-9210-35a6b0c4970d',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'e8a263d7-4230-4bec-92ab-8b80be0000fc',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': 'd766dfe0-ea00-479f-8d1d-82ecb98dd2a9',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'f8ebd592-e64d-4387-8704-ca83c5610e35',\n",
       "  'data_type': 'Raw Intensities',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '3547414c-7f4e-49b4-813f-bc1dbf103fe9',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'data_category': 'Structural Variation',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': '95d72a48-88b5-49a4-80aa-6e39f9821979',\n",
       "  'data_type': 'Masked Intensities',\n",
       "  'data_category': 'DNA Methylation',\n",
       "  'experimental_strategy': 'Methylation Array'},\n",
       " {'uuid': '50e2d5d6-0678-4db7-b849-df7ccd20d424',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'ebf1e95b-4505-420a-9e2b-c552733ceef6',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'data_category': 'Structural Variation',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': '70c28a05-d777-4f2b-943d-2d5d09ad1c20',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'data_category': 'Simple Nucleotide Variation',\n",
       "  'experimental_strategy': 'WXS'},\n",
       " {'uuid': 'dab7a763-f3c8-4532-acb6-1c72f8f29265',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'data_category': 'Structural Variation',\n",
       "  'experimental_strategy': 'RNA-Seq'},\n",
       " {'uuid': 'd05e1fe7-e3c9-4205-abac-c1f57ccddad6',\n",
       "  'data_type': 'Biospecimen Supplement',\n",
       "  'data_category': 'Biospecimen',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'd3480f1c-a301-4799-9d11-afdf81a8bef7',\n",
       "  'data_type': 'Copy Number Segment',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': '3ada67b6-bcdd-4682-b99b-89956219718c',\n",
       "  'data_type': 'miRNA Expression Quantification',\n",
       "  'data_category': 'Transcriptome Profiling',\n",
       "  'experimental_strategy': 'miRNA-Seq'},\n",
       " {'uuid': '234a8d5e-2ad6-4ace-b8fc-e575a8903263',\n",
       "  'data_type': 'Masked Intensities',\n",
       "  'data_category': 'DNA Methylation',\n",
       "  'experimental_strategy': 'Methylation Array'},\n",
       " {'uuid': '49d506d7-09be-4eaa-b1eb-ff7207bd02e3',\n",
       "  'data_type': 'Allele-specific Copy Number Segment',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': 'c78d5190-ac60-4af0-a50a-14183e685b9c',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': 'fa117dd4-b55e-411a-96d1-0b22eb0913af',\n",
       "  'data_type': 'Masked Copy Number Segment',\n",
       "  'data_category': 'Copy Number Variation',\n",
       "  'experimental_strategy': 'Genotyping Array'},\n",
       " {'uuid': 'be7ad45d-05de-4cc8-83d2-3bebaf45d6bd',\n",
       "  'data_type': 'Pathology Report',\n",
       "  'data_category': 'Clinical',\n",
       "  'experimental_strategy': 'N/A'},\n",
       " {'uuid': 'b6aa44b9-8ac4-4b7e-9bee-56e741c7b7fc',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'data_category': 'Sequencing Reads',\n",
       "  'experimental_strategy': 'WGS'}]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_genomic_data_uuids('TCGA-AC-A3W5')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
